File: leetcode/125_valid_palindrome.py
class Solution:

    def isPalindrome(self, s: str) -> bool:
        N = len(s)
        left, right = 0, N - 1
        while True:
            while left < N and not s[left].isalnum():
                left += 1
            while right >= 0 and not s[right].isalnum():
                right -= 1
            if left >= right:
                break
            if s[left].lower() != s[right].lower():
                return False
            left += 1
            right -= 1
        return True

File: leetcode/java/125ValidPalindrome.java
/**
 * T: O(N) S: O(1)
 *
 * <p>Use two pointers, find first alphanumeric charaters from the left and from the right. If they
 * do not match - the string is not a palindrome. Otherwise, proceed to the next pair of
 * alphanumeric characters.
 */
class Solution {
  public boolean isPalindrome(String s) {
    int i = 0;
    int j = s.length() - 1;
    while (i < j) {
      if (!Character.isLetterOrDigit(s.charAt(i))) {
        i++;
        continue;
      }

      if (!Character.isLetterOrDigit(s.charAt(j))) {
        j--;
        continue;
      }
      char left = Character.toLowerCase(s.charAt(i));
      char right = Character.toLowerCase(s.charAt(j));
      if (left != right) {
        return false;
      }
      i++;
      j--;
    }
    return true;
  }
}
